### PR TITLE
0412 use variform for client_attrs

### PR DESCRIPTION
**Pull request by zmstone**
Fixes [EMQX-11882](https://emqx.atlassian.net/browse/EMQX-11882)
This PR rewrites https://github.com/emqx/emqx/pull/12750 which is not released yet.

Release version: v/e5.7.0

## Summary

- Allow multiple `client_attrs` initial extractions
- Support using variform for `client_attrs` initial extraction
- Provide regexp based extraction as a variform function (so the config schema introduced in #12750 is rewritten)

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [x] Changed lines covered in coverage report
- [x] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [x] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

[EMQX-11882]: https://emqx.atlassian.net/browse/EMQX-11882?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by terry-xiaoyu — APPROVED** [review]

File: apps/emqx_utils/src/emqx_variform_bif.erl [hunk 1]
```diff
@@ -47,15 +45,37 @@
...
     unescape/1,
-    nth/2
+    any_to_str/1
 ]).
 
+%% Array functions
+-export([nth/2]).
+
+%% Control functions
+-export([coalesce/1, coalesce/2]).
```
Comment (savonarola) [inline apps/emqx_utils/src/emqx_variform_bif.erl:62]: Wondering: don't we want to make `coalesce` lazy?

Also, a question: how is a customer intended to emulate `IFF(.., .., ...)` (functional if)? Via `coalesce`?

**Review by savonarola — APPROVED** [review]

File: apps/emqx_utils/src/emqx_variform_bif.erl (hunk 1 above)
Comment (zmstone) [reply apps/emqx_utils/src/emqx_variform_bif.erl:62]: > Wondering: don't we want to make coalesce lazy?
Good idea, maybe for improvement. The `eval` loop can be enhanced.

> emulate IFF(cond, a, b)

It's easy: `coalesce([coalesce([a, '']), b])`

Comment (zmstone) [reply apps/emqx_utils/src/emqx_variform_bif.erl:62]: can add a `IFF` function

Comment (zmstone) [reply apps/emqx_utils/src/emqx_variform_bif.erl:62]: To make it more complete, we can:

- `coalesce` should be a reserved word in the expression token scanner
- will need to add `iif` function (reserved word too) for if this else that
- to support `iif`, we'll need more compare functions, such as `gt`, `gte`, `eq`, `lt`, `lte`.
- then there is perhaps to be match functions too

Then if we add variable bindings, it'll become even more complete.

Finally allow user defined functions to be a programing language.

Comment (savonarola) [reply apps/emqx_utils/src/emqx_variform_bif.erl:62]: The idea of the use case that occurred to me while reviewing was checking on emptiness in the `iif` condition.

**Comment by emqxqa** [comment]
[Test Execution]()

**Pull request merged**